runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -147,16 +182,22 @@ private synchronized void removeCompletedWorkFromQueue(
                 () ->
                     new IllegalStateException(
                         String.format(
-                            "Active key %s without work, expected token %d",
-                            shardedKey, workToken)));
+                            "Active key %s without work, expected work_token %d, expected cache_token %d",
+                            shardedKey, workDedupeKey.workToken(), workDedupeKey.cacheToken())));
 
-    if (completedWork.getWorkItem().getWorkToken() != workToken) {
+    if (completedWork.getWorkItem().getWorkToken() != workDedupeKey.workToken()
+        || completedWork.isRetryOf(workDedupeKey)) {
       // Work may have been completed due to clearing of stuck commits.
       LOG.warn(
-          "Unable to complete due to token mismatch for key {} and token {}, actual token was {}.",
+          "Unable to complete due to token mismatch for "
+              + "key {}, work_token {}, and cache_token {}; "
+              + "actual work_token was {},"
+              + "actual cache_token was {}.",
           shardedKey,
-          workToken,
-          completedWork.getWorkItem().getWorkToken());
+          workDedupeKey.workToken(),
```

Comment:
add toString for new class and use it?

Reply:
done auto value has the out of the box so just printing the object